# PARTI <Python程式設計>
## 2. 資料型別與基本運算
- [2.1 數值運算](#數值運算)
- [2.2 字串運算](#字串運算)

<a id='數值運算'></a>
### 2.1 數值運算
- int
- float
- Decimal
- Fraction
- complex number

In [2]:
2+28 

30

In [3]:
150 - 5*26

20

#### Division(/) always return a floating point number.

In [4]:
(150-5*20)/5

10.0

#### floor division(//) return a quotient

In [5]:
79//3

26

##### modulus operator(%) return a remainder

In [6]:
79%3

1

#### operator to calculate power (**)

In [7]:
2**2

4

#### Assignment operator(=)

In [8]:
interest_rate = 0.6
interest_rate

0.6

#### the last printed expression is assigned to the variable (_)

In [9]:
_ +1

1.6

### Decimal objects
- https://docs.python.org/3/library/decimal.html#decimal.Decimal

#### [NOTE]: There are some small differences between arithmetic on Decimal objects and arithmetic on integers and floats. 

In [11]:
from decimal import Decimal
Decimal(-11) % Decimal(4)

Decimal('-3')

### `Fraction` objects 有理數
- https://docs.python.org/3/library/fractions.html#fractions.Fraction

In [13]:
from fractions import Fraction
Fraction(30,-4)

Fraction(-15, 2)

In [15]:
Fraction('-4/23')

Fraction(-4, 23)

#### finite floating number => a rational number = int/int

In [16]:
Fraction(3.33)

Fraction(1874623344892969, 562949953421312)

In [17]:
from decimal import Decimal
Fraction(Decimal('3.33'))    ##  Why?

Fraction(333, 100)

#### Why Fraction(333, 100) ? 
#### Ans : The Fraction class inherits from the abstract base class `numbers.Rational`, and implements all of the methods and operations from that class.

<a id='字串運算'></a>
### 2.2 字串運算

In [18]:
'Hello Python'

'Hello Python'

In [19]:
'isn\'t' # use \' to escape the single quote

"isn't"

In [21]:
"isn\'t" #use double quotes is ok, too

"isn't"

#### printing the string with `print()` function

In [22]:
print('No, it\'s a string')

No, it's a string


In [23]:
print('Hello \n Hi') #\n means new line

Hello 
 Hi


#### use raw strings by adding an `r` before the first quote to avoid the escape character `\` :

In [24]:
print('C:\python\notebook')

C:\python
otebook


In [25]:
print(r'C:\python\notebook')

C:\python\notebook


#### using triple-quotes: `"""..."""` or `'''...'''` to span string literals into multiple lines.

In [28]:
print("""\
[Agendar]
    9:00~9:30   reception
    9:30~10:00  keynote
""")

[Agendar]
    9:00~9:30   reception
    9:30~10:00  keynote



#### Strings' concatenation (+) and repeatition (*):

In [30]:
'repeat' + ' test '*2 + 'end'

'repeat test  test end'

In [34]:
'repea' ('ti'*2) 'on'     ##  'repea' + 'ti'*2 + 'on' 

SyntaxError: invalid syntax (<ipython-input-34-0d17f6c1facc>, line 1)

In [35]:
text = 'Hello Python!' # index start from 0
text[0]

'H'

In [37]:
text[-1] # '-1' mean the last character

'!'

In [38]:
text[1:20]

'ello Python!'

In [39]:
text[:20]

'Hello Python!'

#### get the length of string literal

In [42]:
print(text)
print(text[12])
len(text)

Hello Python!
!


13

#### convert other data type to string type  `str` function

In [43]:
str(3.222222)

'3.222222'